doc/developers/cython.rst

```diff
@@ -0,0 +1,178 @@
+.. _cython:
+
+Cython Best Practices, Conventions and Knowledge
+================================================
+
+This documents:
+
+* technical aspects of Cython which aren't currently present in Cython's documentation
+* tips to develop Cython code in scikit-learn
+
+Cython Internals
+----------------
+
+Concrete method dispatch implementation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+* In most cases, concrete method dispatch is performed **at runtime using virtual tables**
+  (i.e. static C-array of pointers to a given types of function, used at runtime to known which
+  function to call from an instance). This is costly because there is a double indirection.
+
+* In some cases, concrete method dispatch is performed **statically** (a function is called
+  directly by a pointer dereference). This is less costly, especially if there are rooms
+  for this function to be inline.
+
+Known Limitations
+^^^^^^^^^^^^^^^^^
+
+* No template for types exist in Cython, alternatives are:
+
+  * using fused-types
+  * using Tempita, a small preprocessing language to expand Cython source
+
+* Multiple inheritance is not possible for extension types
+* No concept similar to interfaces or traits exist in Cython
+* Currently it is impossible to inherit and override a method which has a fused type in its signature
+* Structs cannot be created using constructor syntax with `nogil` `issue and workaround <https://github.com/cython/cython/issues/1642>`_
+
+Performances improvements will come with `Cython 3.0 which is to be released soon in beta apparently <https://github.com/cython/cython/issues/4022#issuecomment-1340918590>`_
+
+Basics
+^^^^^^
+
+* ``nogil`` declaration: https://github.com/cython/cython/issues/2798#issuecomment-459971828
+
+Tips for developping Cython code in scikit-learn
+------------------------------------------------
+
+Tips to ease development
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+* If you intend to use OpenMP: On MacOS, system's distribution of ``clang`` does not implement OpenMP.
+  You can install the ``compilers`` package available on ``conda-forge`` which comes with implementations of OpenMP.
+
+* Activating `checks <https://github.com/scikit-learn/scikit-learn/blob/62a017efa047e9581ae7df8bbaa62cf4c0544ee4/sklearn/_build_utils/__init__.py#L68-L87>`_ might help. E.g. for activating boundscheck use:
+
+  .. code-block:: bash
+
+         export SKLEARN_ENABLE_DEBUG_CYTHON_DIRECTIVES=1
+
+* Start from scratch in a notebook to understand.
```

Comment:
You can give some details here:
- first load cython extension ``%load_ext cython``
- to run a cython cell ``%%cython``
- some useful flags
    - ``%%cython -f`` to force recompiling each time you run the cell
    - ``%%cython --compile-args=-fopenmp`` if trouble compiling openmp code (replace by the appropriate flag for the compiler)
- maybe even link to the cython magic doc https://ipython.org/ipython-doc/2/config/extensions/cythonmagic.html

Reply:
I think I better link to [this part of the documentation](https://cython.readthedocs.io/en/latest/src/quickstart/build.html#using-the-jupyter-notebook). Then we can contribute your proposed elements to Cython's documentation directly.

What do you think? :slightly_smiling_face:

Reply:
I agree, better link to the official documentation.
I still think that it's worth mentioning ``--compile-args=-fopenmp`` if trouble compiling openmp code (replace by the appropriate flag for the compiler) because it's not in the linked page.